treasuries/src/lib.rs

```diff
@@ -17,47 +17,129 @@
 #![cfg_attr(not(feature = "std"), no_std)]
 
 use core::marker::PhantomData;
-use encointer_primitives::communities::CommunityIdentifier;
+use encointer_primitives::{balances::BalanceType, communities::CommunityIdentifier};
 use frame_support::{
 	traits::{Currency, ExistenceRequirement::KeepAlive, Get},
 	PalletId,
 };
+use frame_system::ensure_signed;
 use log::info;
 use parity_scale_codec::Decode;
 use sp_core::H256;
 use sp_runtime::traits::Hash;
-
 // Logger target
 const LOG: &str = "encointer";
 
+pub use crate::weights::WeightInfo;
 pub use pallet::*;
 
 #[cfg(test)]
 mod mock;
 #[cfg(test)]
 mod tests;
+mod weights;
 
 pub type BalanceOf<T> =
 	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
 
 #[frame_support::pallet]
 pub mod pallet {
 	use super::*;
+	use encointer_primitives::treasuries::SwapNativeOption;
 	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::OriginFor;
 
 	#[pallet::pallet]
 	pub struct Pallet<T>(PhantomData<T>);
 
 	#[pallet::config]
-	pub trait Config: frame_system::Config {
+	pub trait Config:
+		frame_system::Config + pallet_encointer_balances::Config + pallet_timestamp::Config
+	{
 		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
 		type Currency: Currency<Self::AccountId>;
 
 		/// The treasuries' pallet id, used for deriving sovereign account IDs per community.
 		#[pallet::constant]
 		type PalletId: Get<PalletId>;
+
+		// /// the maximum fraction of available treasury funds a single swap can claim
+		// /// defined as divisor: 2 means half of the available funds can be swapped
+		// #[pallet::constant]
+		// type MaxFractionPerSwap: Get<u8>;
+		//
+		// /// the minimum period an account has to wait between two swaps
+		// #[pallet::constant]
+		// type SwapCooldownPeriod: Get<T::Moment>;
+
+		type WeightInfo: WeightInfo;
 	}
 
+	#[pallet::storage]
+	#[pallet::getter(fn swap_native_options)]
+	pub type SwapNativeOptions<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		CommunityIdentifier,
+		Blake2_128Concat,
+		T::AccountId,
+		SwapNativeOption<BalanceOf<T>, T::Moment>,
+		OptionQuery,
+	>;
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		sp_core::H256: From<<T as frame_system::Config>::Hash>,
+		T::AccountId: AsRef<[u8; 32]>,
+		u64: TryFrom<BalanceOf<T>>,
```

Comment:
Those trait bounds seem a bit hacky. Additionally, I just double checked; you can simply remove the `T::AccountId: AsRef<[u8; 32]>` bound.

Reply:
You can also remove the u64, with some trickery, see my PR into your PR: https://github.com/encointer/pallets/pull/413